daemon/daemon.go

```diff
@@ -1032,23 +1036,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 	return nil
 }
 
+func writeDistributionProgress(cancelFunc func(), outStream io.Writer, progressChan <-chan xfer.Progress) {
+	sf := streamformatter.NewJSONStreamFormatter()
+	for prog := range progressChan {
+		if prog.Message != "" {
+			formatted := sf.FormatStatus(prog.ID, prog.Message)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
+			}
+		} else {
+			jsonProgress := jsonmessage.JSONProgress{Current: prog.Current, Total: prog.Total}
+			formatted := sf.FormatProgress(prog.ID, prog.Action, &jsonProgress)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
+			}
+		}
+	}
+}
+
 // PullImage initiates a pull operation. image is the repository name to pull, and
 // tag may be either empty, or indicate a specific tag to pull.
 func (daemon *Daemon) PullImage(ref reference.Named, metaHeaders map[string][]string, authConfig *cliconfig.AuthConfig, outStream io.Writer) error {
+	// Include a buffer so that slow client connections don't affect
+	// transfer performance.
+	progressChan := make(chan xfer.Progress, 100)
+
+	writesDone := make(chan struct{})
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	go func() {
+		writeDistributionProgress(cancelFunc, outStream, progressChan)
+		close(writesDone)
+	}()
+
 	imagePullConfig := &distribution.ImagePullConfig{
 		MetaHeaders:     metaHeaders,
 		AuthConfig:      authConfig,
-		OutStream:       outStream,
+		ProgressChan:    progressChan,
 		RegistryService: daemon.RegistryService,
 		EventsService:   daemon.EventsService,
 		MetadataStore:   daemon.distributionMetadataStore,
-		LayerStore:      daemon.layerStore,
 		ImageStore:      daemon.imageStore,
 		TagStore:        daemon.tagStore,
-		Pool:            daemon.distributionPool,
+		DownloadManager: daemon.downloadManager,
 	}
 
-	return distribution.Pull(ref, imagePullConfig)
+	err := distribution.Pull(ctx, ref, imagePullConfig)
```

Comment:
I can see the appeal of doing this, but I'm reluctant to store objects on contexts in general. I don't like how it loses type safety and needs to be wrapped with special utility functions to do the proper type assertions. I also don't like the way variables are looked up by string names, adding overhead and making name conflicts possible. Contexts just remind me too much of Python :)

What do other people think?

`ImagePullConfig` definitely needs to be cleaned up. This could either happen as part of this PR, or as a followup. I'm thinking the daemon could create a `Distribution` object at startup that references static things like the image store and layer store. `Pull` and `Push` could be methods on that object.